### PR TITLE
Implement `chdir`, `fchdir`, `getcwd` and `ttyname`

### PR DESCRIPTION
**Pull request by Urgau**
This pull-request implement the following functions: `chdir`, `fchdir`, `getcwd` and `ttyname` with their unit test (expect for `ttyname`, because I don't know how to test it reliably and simply).

`getcwd` and `ttyname` use the same routine that was already present for reusing an `OsString` if possible.

### CODE REVIEW

File: src/imp/linux_raw/syscalls.rs [hunk 1]
```diff
@@ -3012,6 +3017,56 @@ pub(crate) fn gettid() -> Pid {
...
+
+    // Get the proc path of the tty
+    let procname = procfdname(fd.as_raw_fd());
+
+    // And finaly gatter the ttyname by reading the proc path
+    readlink(
+        // SATEFY: procfdname always return a nul terminated string
+        unsafe { CStr::from_bytes_with_nul_unchecked(&procname) },
+        buf,
+    )?;
```
Comment (sunfishcode) [inline src/imp/linux_raw/syscalls.rs:3065]: `readlink("/proc/self/fd/...")` isn't always reliable; file handles for deleted files or pipes end up being symlinks to special strings which aren't paths. We should check that the resulting path can be opened with `stat` and that the result's `st_dev`+`st_ino` matches that of `fd`.

File: src/imp/linux_raw/syscalls.rs [hunk 2]
```diff
@@ -3012,6 +3017,56 @@ pub(crate) fn gettid() -> Pid {
...
+    buf
+}
+
+#[inline]
+pub(crate) fn ttyname(fd: BorrowedFd<'_>, buf: &mut [u8]) -> io::Result<()> {
+    // Check that the fd is really a tty
+    ioctl_tiocgwinsz(fd)?;
+
+    // Get the proc path of the tty
+    let procname = procfdname(fd.as_raw_fd());
```
[the remark this answers is not in the document]
Comment (Urgau) [reply src/imp/linux_raw/syscalls.rs]: I should have check more carefully :grinning:. Done.

File: src/fs/cwd.rs [hunk 3]
```diff
@@ -27,3 +29,37 @@ pub fn cwd() -> BorrowedFd<'static> {
         BorrowedFd::<'static>::borrow_raw_fd(at_fdcwd)
     }
 }
+
+/// `getcwd(reuse)`
```
[the remark this answers is not in the document]
Comment (Urgau) [reply src/fs/cwd.rs]: Done. I was indeed a bit confuse about that, it's the reason why I wrote `reuse` but didn't seem right to me. Not writing anything can have a meaning. I don't know what would be better; in the mean time a removes `reuse` part.

File: src/imp/linux_raw/syscalls.rs (hunk 1 above)
Comment (Urgau) [reply src/imp/linux_raw/syscalls.rs:3041]: Yeah you're right, I was a little bit lazy after 25 github actions to get every `cfg` right. Done.

Comment (sunfishcode) [reply src/imp/linux_raw/syscalls.rs:3041]: I can sympathize. It's a lot :-}.

**Comment by sunfishcode** [comment]
Looks good!

**Pull request merged**